samples/scan/scan.cpp

```diff
@@ -57,104 +57,210 @@ struct identity<T, std::logical_and<T>> {
+  size_t global_mem_size = dev.get_info<sycl::info::device::global_mem_size>();
+  if (in.get_count() > (global_mem_size / 2)) {
+    throw std::runtime_error("Input size exceeds device global memory size.");
+  }
 
-            /* Perform down-sweep on the tree to compute the whole scan.
-             * Again, `off` is 2^d. */
-            for (size_t off = size / 2; off > 0; off >>= 1) {
-              item.barrier(sycl::access::fence_space::local_space);
-
-              size_t i = gid * off * 2;
-              if (i < size) {
-                auto t = temp[i + off - 1];
-                auto u = temp[i + off * 2 - 1];
-                temp[i + off - 1] = u;
-                temp[i + off * 2 - 1] = Op{}(t, u);
-              }
-            }
+  // Check if local memory is available.
+  if (dev.get_info<sycl::info::device::local_mem_type>() ==
+      sycl::info::local_mem_type::none) {
+    throw std::runtime_error("Device does not have local memory.");
+  }
 
-            // Synchronize again to observe results.
-            item.barrier(sycl::access::fence_space::local_space);
+  // Obtain device limits.
+  size_t max_wgroup_size =
+      dev.get_info<sycl::info::device::max_work_group_size>();
+  size_t local_mem_size = dev.get_info<sycl::info::device::local_mem_size>();
+
+  /* Find a work-group size that is guaranteed to fit in local memory and is
+   * below the maximum work-group size of the device. */
+  size_t wgroup_size_lim =
+      sycl::min(max_wgroup_size, local_mem_size / (2 * sizeof(T)));
+
+  /* Every work-item processes two elements, so the work-group size has to
+   * divide this number evenly. */
+  size_t half_in_size = in.get_count() / 2;
+
+  size_t wgroup_size = 0;
+  /* Find the largest power of two that divides half_in_size and is within the
+   * device limit. */
+  for (size_t pow = size_t(1) << (sizeof(size_t) * 8 - 1); pow > 0; pow >>= 1) {
+    if ((half_in_size / pow) * pow == half_in_size && pow <= wgroup_size_lim) {
+      wgroup_size = pow;
+      break;
+    }
+  }
 
-            /* To return an inclusive rather than exclusive scan result, shift
-             * each element left by 1 when writing back into global memory. */
-            data[2 * gid] = temp[2 * gid + 1];
-            data[2 * gid + 1] = temp[2 * gid + 2];
-          });
-    });
+  if (wgroup_size == 0) {
+    throw std::runtime_error(
+        "Could not find an appropriate work-group size for the given input.");
   }
 
-  // Compute value of last scan which was not included in exclusive scan.
-  out[size - 1] = Op{}(out[size - 2], in[size - 1]);
+  q.submit([&](sycl::handler& cgh) {
+    auto data = in.template get_access<sycl::access::mode::read_write>(cgh);
+    sycl::accessor<T, 1, sycl::access::mode::read_write,
+                   sycl::access::target::local>
 
-  return out;
-}
+        temp(wgroup_size * 2, cgh);
+
+    // Use dummy struct as the unique kernel name.
+    cgh.parallel_for<kernel_name<T, Op, 0>>(
```

Comment:
Hmmm, you could do a fun little trick here by making this in itself a `typename` in `kernel_name`'s declaration, that way you could call it "class segmented_scan", then the next like "class store_largest" or something, to help with readability (this would also come up in debugging). Consider it minor, though - and it's slightly unpleasant

Reply:
Compute++ disallows using nested classes/structs/etc. as kernel names, and a typedef still resolves to the `kernel_name` class, so I left it as is.